### ISSUE TITLE
Elasticluster Fails to Connect to GCE Instances Through SSH

### ISSUE THREAD

**Issue by hheydari** [gc3-uzh-ch/elasticluster#359]
Hi - I'm trying to setup a simple 3-node gridengine on Google Compute using the latest release (cloned from github a day ago) with the following config:

`# Gridengine software to be configured by Ansible
[setup/ansible-gridengine]
provider=ansible
frontend_groups=gridengine_master
compute_groups=gridengine_clients

# Create a cloud provider (call it "google-cloud")
[cloud/google-cloud]
provider=google
gce_project_id=****THIS IS PROJECT ID****
noauth_local_webserver=True
gce_client_id=****MY GCE CLIENT ID****
gce_client_secret=****MY GCE CLIENT SECRET****

# Create a login (call it "google-login")
[login/google-login]
image_user=****THIS IS MY GOOGLE USER ID****
image_user_sudo=root
image_sudo=True
user_key_name=elasticluster
user_key_private=****PRIVATE SSH KEY WITH ITS USER BEING MY GOOGLE USER ID****
user_key_public=****PUBLIC SSH KEY WITH ITS USER BEING MY GOOGLE USER ID****

# Bring all of the elements together to define a cluster called "gridengine"
[cluster/gridengine]
cloud=google-cloud
login=google-login
setup_provider=ansible-gridengine
security_group=default
image_id=ubuntu-1404-trusty-v20161130
flavor=n1-standard-1
frontend_nodes=1
compute_nodes=2
image_userdata=
ssh_to=frontend`

Instances are created and elasticluster recognizes their start. But, it stuck and finally fails after timeout at:
`Checking SSH connection to nodes ...`

Obviously there is something wrong with ssh keys, etc. I tried logging in directly to front end using ssh by using my key and google user id and it works. Perhaps, there is something I forgot to set in the config that elasticluster can't log in?

Could it be related to known_host file? I see that keys are pulled from that in the code. Saying this because when elasticluster is still in the loop trying to connect, I can connect in another terminal and the fingerprint and server haven't been added to known hosts. So I just type yes and they are added but elasticluster still is in the loop and finally fails after a while.

Thanks

**Comment by hheydari** [comment]
Debugging the code, found that at first the SSH connection fails because:

```
Host compute001 (104.196.163.0) not reachable: timed out.
Host frontend001 (104.196.11.15) not reachable: timed out.
etc.
```

But then, the error changes to:

```
Trying to connect to host compute001 (104.196.163.0)
Ignoring error not a valid EC private key file connecting to compute001

Trying to connect to host frontend001 (104.196.11.15)
Ignoring error not a valid EC private key file connecting to frontend001
```

Yet, elasticluster ssh gridengine works!

**Comment by hheydari** [comment]
I wonder if this has something to deal with setting public key in Google cloud console? In meta-data? I've done that.

**Comment by hheydari** [comment]
ok, so I keep adding info here:

Arguments passed to ssh.connect here https://github.com/gc3-uzh-ch/elasticluster/blob/master/elasticluster/cluster.py#L1242 is perfectly fine as far I know. Looks like:

Address: 104.196.163.0
Username:    ***MY GOOGLE ID****
key_filename:   ***PATH TO MY PRIVATE KEY***
timeout: 5
port: 22

**Comment by riccardomurri** [comment]
Hello Hamed,

Thanks for giving this debug information -- it's indeed very useful!

> But then, the error changes to:
> 
> ```
> Trying to connect to host compute001 (104.196.163.0)
> Ignoring error not a valid EC private key file connecting to
> compute001

I think this is another instance of
https://github.com/uzh/gc3pie/issues/589
which see for a more detailed explanation.

(ElastiCluster and GC3Pie share a lot of code in their cloud backends,
so it's quite possible that bugs are shared too.)

In short, you would hit this bug if all the following three conditions
are met:

1. You are not using `ssh-agent`, or the SSH key used to connect to the
   VMs is not loaded into the agent.
2. The SSH private key is protected with a passphrase
3. The SSH private key has a non-default name (e.g., it's an RSA key but
   it's not named `id_rsa`)

Does this describe your current set up?


> Yet, elasticluster ssh gridengine works!

Yes: the above bug stems from (ab)use of Paramiko (Python's SSH
library), whereas `elasticluster ssh` is just a front-end for the
OpenSSH `ssh` command which is implemented in C and thus uses a
completely different code path and key handling library.

Hope this helps,
Riccardo

**Comment by hheydari** [comment]
@riccardomurri Yes, that's it! I totally forgot adding the key to the agent for that terminal session. I guess I should mark this as solved, but maybe you want to add some check in the code for this situation? Up to you to me it's solved and what a lifesaver elasticluster is! Thank you and all other contributors...

**Comment by hheydari** [comment]
Oh, by the way, GC3Pie is awesome! You really need to advertise it in stackoverflow, etc. by answering questions related to Python dealing with job submission to SGE, etc. 

This is great! I'm glad I didn't add the key to the ssh session as I found about GC3Pie 👍

**Comment by riccardomurri** [comment]
@hheydarii Yes, please leave the issue report open so I'll remember to fix the underlying code to produce a meaningful error message.

And thanks for the kind words about GC3Pie, I'm glad you like it!
